File: src-tauri/src/commands/likes.rs
//! Tauri commands for post likes/reactions

use crate::db::repositories::{LikeSummary, LikesRepository};
use crate::db::Database;
use crate::error::{AppError, Result};
use crate::services::{IdentityService, WallSocialService};
use std::sync::Arc;
use tauri::State;

/// Like a post
#[tauri::command]
pub async fn like_post(
    db: State<'_, Arc<Database>>,
    identity_service: State<'_, Arc<IdentityService>>,
    wall_social_service: State<'_, Arc<WallSocialService>>,
    post_id: String,
) -> Result<LikeSummary> {
    let identity = identity_service
        .get_identity()?
        .ok_or_else(|| AppError::IdentityNotFound("No identity found".to_string()))?;
    wall_social_service.add_reaction(&post_id, "like")?;
    LikesRepository::get_like_summary(&db, &post_id, &identity.peer_id)
        .map_err(|e| AppError::DatabaseString(e.to_string()))
}

/// Unlike a post
#[tauri::command]
pub async fn unlike_post(
    db: State<'_, Arc<Database>>,
    identity_service: State<'_, Arc<IdentityService>>,
    wall_social_service: State<'_, Arc<WallSocialService>>,
    post_id: String,
) -> Result<LikeSummary> {
    let identity = identity_service
        .get_identity()?
        .ok_or_else(|| AppError::IdentityNotFound("No identity found".to_string()))?;
    wall_social_service.remove_reaction(&post_id, "like")?;
    LikesRepository::get_like_summary(&db, &post_id, &identity.peer_id)
        .map_err(|e| AppError::DatabaseString(e.to_string()))
}

/// Get like summary for a single post
#[tauri::command]
pub async fn get_post_likes(
    db: State<'_, Arc<Database>>,
    identity_service: State<'_, Arc<IdentityService>>,
    post_id: String,
) -> Result<LikeSummary> {
    // Get current identity (or use empty string for non-logged-in users)
    let current_peer_id = identity_service
        .get_identity()?
        .map(|i| i.peer_id)
        .unwrap_or_default();

    LikesRepository::get_like_summary(&db, &post_id, &current_peer_id)
        .map_err(|e| AppError::DatabaseString(e.to_string()))
}

/// Get like summaries for multiple posts (efficient batch query)
#[tauri::command]
pub async fn get_posts_likes_batch(
    db: State<'_, Arc<Database>>,
    identity_service: State<'_, Arc<IdentityService>>,
    post_ids: Vec<String>,
) -> Result<Vec<LikeSummary>> {
    // Get current identity (or use empty string for non-logged-in users)
    let current_peer_id = identity_service
        .get_identity()?
        .map(|i| i.peer_id)
        .unwrap_or_default();

    LikesRepository::get_like_summaries_batch(&db, &post_ids, &current_peer_id)
        .map_err(|e| AppError::DatabaseString(e.to_string()))
}

/// Get all posts that the current user has liked
#[tauri::command]
pub async fn get_my_liked_posts(
    db: State<'_, Arc<Database>>,
    identity_service: State<'_, Arc<IdentityService>>,
) -> Result<Vec<String>> {
    // Get current identity
    let identity = identity_service
        .get_identity()?
        .ok_or_else(|| AppError::IdentityNotFound("No identity found".to_string()))?;

    LikesRepository::get_liked_posts(&db, &identity.peer_id)
        .map_err(|e| AppError::DatabaseString(e.to_string()))
}

File: src/services/likes.ts
import { invokeCommand } from './command';

/** Summary of like/reaction state for a post. */
export interface LikeSummary {
  postId: string;
  totalLikes: number;
  userHasLiked: boolean;
}

interface RawLikeSummary {
  postId?: string;
  post_id?: string;
  totalLikes?: number;
  total_likes?: number;
  userHasLiked?: boolean;
  user_has_liked?: boolean;
}

function normalizeLikeSummary(summary: RawLikeSummary): LikeSummary {
  return {
    postId: summary.postId ?? summary.post_id ?? '',
    totalLikes: summary.totalLikes ?? summary.total_likes ?? 0,
    userHasLiked: summary.userHasLiked ?? summary.user_has_liked ?? false,
  };
}

/** Likes service - wraps Tauri commands for signed post reactions. */
export const likesService = {
  async likePost(postId: string): Promise<LikeSummary> {
    return normalizeLikeSummary(await invokeCommand('like_post', { postId }));
  },

  async unlikePost(postId: string): Promise<LikeSummary> {
    return normalizeLikeSummary(await invokeCommand('unlike_post', { postId }));
  },

  async getPostLikes(postId: string): Promise<LikeSummary> {
    return normalizeLikeSummary(await invokeCommand('get_post_likes', { postId }));
  },

  async getPostsLikesBatch(postIds: string[]): Promise<LikeSummary[]> {
    const summaries = await invokeCommand('get_posts_likes_batch', { postIds });
    return summaries.map(normalizeLikeSummary);
  },

  async getMyLikedPosts(): Promise<string[]> {
    return invokeCommand('get_my_liked_posts');
  },
};
